ksp_plugin_test/orbit_analyser_test.cpp

```diff
@@ -0,0 +1,137 @@
+using ::testing::IsNull;
+using ::testing::Optional;
+using ::testing::Property;
+
+class OrbitAnalyserTest : public testing::Test {
+ protected:
+  OrbitAnalyserTest()
+      : earth_1957_(RemoveAllButEarth(SolarSystem<Barycentric>(
+            SOLUTION_DIR / "astronomy" / "sol_gravity_model.proto.txt",
+            SOLUTION_DIR / "astronomy" /
+                "sol_initial_state_jd_2436116_311504629.proto.txt",
+            /*ignore_frame=*/true))),
+        // As there is only one body left in this ephemeris, integration is
+        // exact up to roundoff error regardless of the step size.  Use a long
+        // step so that we do not waste time computing the ephemeris.
+        ephemeris_(earth_1957_.MakeEphemeris(
+            /*accuracy_parameters=*/{/*fitting_tolerance=*/1 * Milli(Metre),
+                                     /*geopotential_tolerance=*/0x1p-24},
+            Ephemeris<Barycentric>::FixedStepParameters(
+                SymmetricLinearMultistepIntegrator<QuinlanTremaine1990Order12,
+                                                   Position<Barycentric>>(),
+                /*step=*/1 * JulianYear))),
+        earth_(*earth_1957_.rotating_body(*ephemeris_, "Earth")),
+        itrs_(ephemeris_.get(), &earth_),
+        topex_poséidon_(SOLUTION_DIR / "astronomy" / "standard_product_3" /
+                            "grgtop03.b97344.e97348.D_S.sp3",
+                        StandardProduct3::Dialect::GRGS) {}
+
+  SolarSystem<Barycentric> earth_1957_;
+  not_null<std::unique_ptr<Ephemeris<Barycentric>>> ephemeris_;
+  RotatingBody<Barycentric> const& earth_;
+  BodySurfaceDynamicFrame<Barycentric, ITRS> itrs_;
+  StandardProduct3 topex_poséidon_;
+
+ private:
+  SolarSystem<Barycentric> RemoveAllButEarth(
+      SolarSystem<Barycentric> solar_system) {
+    std::vector<std::string> const names = solar_system.names();
+    for (auto const& name : names) {
+      if (name != "Earth") {
+        solar_system.RemoveMassiveBody(name);
+      }
+    }
+    return solar_system;
+  }
+};
+
+TEST_F(OrbitAnalyserTest, TOPEXPoséidon) {
+  OrbitAnalyser analyser(ephemeris_.get(), DefaultHistoryParameters());
+  EXPECT_THAT(analyser.analysis(), IsNull());
+  EXPECT_THAT(analyser.next_analysis_percentage(), Eq(0));
+  auto const& arc =
+      *topex_poséidon_.orbit(
+          {StandardProduct3::SatelliteGroup::General, 1}).front();
+  ephemeris_->Prolong(arc.begin()->time);
+  analyser.RequestAnalysis(arc.begin()->time,
+                           itrs_.FromThisFrameAtTime(arc.begin()->time)(
+                               arc.begin()->degrees_of_freedom),
+                           3 * Hour,
+                           &earth_);
+  while (analyser.next_analysis_percentage() != 100) {
+    absl::SleepFor(absl::Milliseconds(10));
+  }
+  // Since the progress percentage only tracks the integration, not the
+  // analysis, we have no guarantee that an analysis is available immediately.
+  do {
+    analyser.RefreshAnalysis();
```

Comment:
So the previous loop could be subsumed by this one?

Also, sleep in the loop, no busy waiting.

Reply:
It could, but by having the previous loop I also test that the analysis percentage reaches 100.